app/components/Form/index.js

```diff
@@ -0,0 +1,271 @@
+    };
+  };
+
+  setFormState(newFormState) {
+    // set form state
+    this.formState = newFormState;
+
+    // get form data and set it in the broadcast
+    const formData = this.getFormData();
+    this.formDataBroadcast.setState(formData);
+
+    // if a handler exists, call it with form data
+    if (this.props.onFormDataChange) {
+      this.props.onFormDataChange(formData);
+    }
+  }
+
+  /**
+   * Mark every field as blurred.
+   *
+   * This is called when the user tries to submit the form so that error messages
+   * are visible for every field.
+   */
+  blurAll() {
+    const { fields } = this.props;
+
+    const allBlurred = fields.reduce((accumulator, field) => {
+      accumulator[field] = true;
+      return accumulator;
+    }, {});
+
+    this.setFormState({
+      ...this.formState,
+      blurred: allBlurred,
+    });
+  }
+
+  /**
+   * Handle blurring a form field.
+   *
+   * Sets `blurred` to true for the field.
+   */
+  handleBlur = fieldName => {
+    this.setFormState({
+      ...this.formState,
+      blurred: {
+        ...this.formState.blurred,
+        [fieldName]: true,
+      },
+    });
+  };
+
+  /**
+   * Handle changing a form field.
+   *
+   * Sets the new field value.
+   */
+  handleChange = (fieldName, value) => {
+    this.setFormState({
+      ...this.formState,
+      values: {
+        ...this.formState.values,
+        [fieldName]: value,
+      },
+    });
+  };
+
+  /**
+   * Initialize form state with the provided form values
+   */
+  initializeForm = newValues => {
+    const initialFormState = this.getInitialFormState(newValues);
+    this.setFormState(initialFormState);
+  };
+
+  /**
+   * Handle submitting the form.
+   *
+   * Checks if the form is valid, and then calls the provided `onSubmit` function.
+   * Returns true if the submit was valid.
+   */
+  handleSubmit = () => {
+    const { onSubmit } = this.props;
+    const { values } = this.formState;
+
+    const errors = this.runValidator();
+    const isFormValid = Object.keys(errors).length === 0;
+
+    // mark all fields as blurred, that way errors can be shown in the UI
+    this.blurAll();
+
+    if (isFormValid && onSubmit) {
+      onSubmit(values);
+      return true;
+    }
+    return false;
+  };
+
+  /**
+   * Runs the provided `validator`.
+   *
+   * Falls back to a default validator which always returns valid.
+   *
+   * (values, props) => {ErrorObject}
+   * A validator takes the current values and props, and returns an object
+   * mapping every failing field to a string describing the error.
+   *
+   * A special `_error` key represents a generic error in the form.
+   */
+  runValidator() {
+    const { validate } = this.props;
+    const { values } = this.formState;
+
+    const defaultValidator = () => ({});
+    const validator = validate || defaultValidator;
+
+    return validator(values, this.getOtherProps());
+  }
+
+  render() {
+    return this.props.children;
```

Comment:
I feel there should be a `<form onSubmit={this.onSubmit}>` (with preventDefault). It will help with submissions on Enter.

Reply:
Done. Here's the changes:

- Now `<Form>` renders a `<form>` HTML element, with `onSubmit`, `preventDefault` and "submit on enter" handled automatically.
- EntryFormRow was updated to render the form within a `<td>` element to be correct HTML.